frontend/src/lib/OutputArtifactLoader.ts

```diff
@@ -198,6 +217,278 @@ export class OutputArtifactLoader {
+        arguments: JSON.stringify(specifiedArguments),
+        source: '',
+        type: ApiVisualizationType.CUSTOM,
+      };
+      const visualization = await Apis.buildPythonVisualizationConfig(visualizationData);
+      if (!visualization.htmlContent) {
+        throw new Error('Failed to build TFDV artifact visualization');
+      }
+      return {
+        htmlContent: visualization.htmlContent,
+        type: PlotType.WEB_APP,
+      } as HTMLViewerConfig;
+    });
+  }
+
+  public static filterTfmaArtifactsPaths(
+    artifactTypes: ArtifactType[],
+    artifacts: Artifact[],
+  ): string[] {
+    const tfmaArtifactTypes = artifactTypes.filter(artifactType => {
+      return artifactType.getName() === 'ModelEvaluation';
+    });
+    const tfmaArtifactTypeIds = tfmaArtifactTypes.map(artifactType => {
+      return artifactType.getId();
+    });
+    const tfmaArtifacts = artifacts.filter(artifact => {
+      return tfmaArtifactTypeIds.includes(artifact.getTypeId());
+    });
+
+    const tfmaArtifactPaths: string[] = [];
+    tfmaArtifacts.forEach(artifact => {
+      const uri = artifact.getUri();
+      if (uri && uri.length > 0) {
+        tfmaArtifactPaths.push(uri);
+      }
+    });
+    return tfmaArtifactPaths;
+  }
+
+  public static getTfmaArtifactViewers(
+    tfmaArtifactPaths: string[],
+  ): Array<Promise<HTMLViewerConfig>> {
+    return tfmaArtifactPaths.map(async artifactPath => {
+      const script = [
+        'import tensorflow_model_analysis as tfma',
+        "tfma_result = tfma.load_eval_result('" + artifactPath + "')",
+        'tfma.view.render_slicing_metrics(tfma_result)',
+      ];
+      const specifiedArguments: any = JSON.parse('{}');
+      specifiedArguments.code = script;
+      const visualizationData: ApiVisualization = {
+        arguments: JSON.stringify(specifiedArguments),
+        source: '',
+        type: ApiVisualizationType.CUSTOM,
+      };
+      const visualization = await Apis.buildPythonVisualizationConfig(visualizationData);
+      if (!visualization.htmlContent) {
+        throw new Error('Failed to build TFMA artifact visualization');
+      }
+      return {
+        htmlContent: visualization.htmlContent,
+        type: PlotType.WEB_APP,
+      } as HTMLViewerConfig;
+    });
+  }
+
+  public static async buildTFXArtifactViewerConfig(
+    kfpPodName: string,
+  ): Promise<HTMLViewerConfig[]> {
+    // Since artifact types don't change per run, this can be optimized further so
+    // that we don't fetch them on every page load.
+    const artifactTypes = await this.getArtifactTypes();
+    if (!artifactTypes) {
+      logger.error('Failed getting artifact types');
+      return [];
+    }
+
+    const context = await this.getMlmdContext(kfpPodName);
+    if (!context) {
+      logger.error('Failed finding corresponding MLMD context');
+      return [];
+    }
+
+    const execution = await this.getExecutionInContextWithPodName(kfpPodName, context);
```

Comment:
Is context really needed to retrieving the Execution?

Reply:
For now, yes. I believe MLMD may add APIs in the future to query executions with particular properties, but until then, we have to get the context first (please let me know if you're aware of any other solutions though!)